### PR TITLE
Make cache accept callable arguments, add tests.

### PR DESCRIPTION
**Pull request by grundic**
Hi, I've added possibility to use cache as callable and added some tests.

### CODE REVIEW

File: boltons/cacheutils.py [hunk 1]
```diff
@@ -450,7 +450,7 @@ def _make_cache_key(args, kwargs, typed=False, kwarg_mark=_KWARG_MARK,
 class CachedFunction(object):
     def __init__(self, func, cache, typed=False):
         self.func = func
-        self.cache = cache
+        self.cache = cache() if hasattr(cache, '__call__') else cache
```
Comment (mahmoud) [inline boltons/cacheutils.py]: this should almost certainly be using [callable](https://docs.python.org/2/library/functions.html#callable)

File: tests/test_cacheutils.py [hunk 2]
```diff
@@ -2,7 +2,9 @@
 
 import string
 
-from boltons.cacheutils import LRU, LRI
+import mock
```
Comment (mahmoud) [inline tests/test_cacheutils.py]: Tests look pretty good, but the usage of mock seems excessive. We don't need mock functions, a normal function returning a constant should do just fine.

**Comment by kdeldycke** [comment]
Is this PR covers #66 too?

**Comment by mahmoud** [comment]
@kdeldycke no, not at all :stuck_out_tongue: soon though!

@grundic upon further review, don't you think this would make more sense in the `__call__`?

**Comment by grundic** [comment]
@mahmoud, yep definitely checking for `callable` should be done via your proposal, fixed that.
What about `mock` - sure I can create simple nested functions and test them, but how do I check call count? It could look something like [this](http://stackoverflow.com/questions/21716940/is-there-a-way-to-track-the-number-of-times-a-function-is-called), but it would make tests a little bit complicated how do you think?

**Comment by mahmoud** [comment]
@grundic My more-important point is that I'm not sure it makes sense to call the cache argument in the `__init__`. If you do that, the user might as well just call it before they pass it in. Instead, the callable should be called in the `cached` decorator's `__call__`.

Still not sure if that sort of lazy access is a good pattern, but it's probably not harmful. I would guess there are precious few callable cache instances that don't return caches.

**Comment by grundic** [comment]
If cache as a function would be in `__call__` it would executed each time the cache is requested. Or there should be some check whether we already executed and getted the cache. In `__init__` we just initializing the value one time.

I don't think that users could easily call the function and pass the value to the decorator, as it evaluated at class/method creation. So it would not be possible to dynamically configure cache at runtime. With function as argument it is possible. Actually it was my naive approach to implement `cachedmethod` pattern with access to cache with `operator`:
```python
from boltons.cacheutils import LRU,cached
import operator

class ClassWithCache(object):
    def __init__(self):
        self.cache = LRU()

    @cached(operator.attrgetter('cache'))
    def number(self, value):
        return value

instance = ClassWithCache()
print(instance.number(19))
```

But as I discovered, it requires a lot more implementation, so I created #70 :)

**Comment by mahmoud** [comment]
I figured it would be easier to show than tell, so I went ahead and implemented this. [Have a look](https://github.com/mahmoud/boltons/blob/c74252986b1fbf89d651f1156127c4077d3aa3ef/boltons/cacheutils.py#L464), maybe take it for a spin, let me know what you think!

**Comment by grundic** [comment]
Excellent, closing my PR.
Thank you!

**Pull request closed without merging**